### PR TITLE
storage: Add support for gcs for Lightning

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve? 

gcs is not supported yet.

### What is changed and how it works?

Add support for gcs.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Related changes

 - Need to cherry-pick to the release branch

### Release note

 -No release note

### CODE REVIEW

File: pkg/storage/gcs.go [hunk 1]
```diff
@@ -258,3 +291,76 @@ func hasSSTFiles(ctx context.Context, bucket *storage.BucketHandle, prefix strin
...
+//
+// Currently, tidb-lightning depends on this method to read parquet file for gcs storage.
+func (r *gcsObjectReader) Seek(offset int64, whence int) (int64, error) {
+	var realOffset int64
+	switch whence {
+	case io.SeekStart:
+		realOffset = offset
+	case io.SeekCurrent:
+		realOffset = r.pos + offset
+	case io.SeekEnd:
```
Comment (Leavrth) [inline pkg/storage/gcs.go:345]: I'm not sure if this means reverse offset. (e.g. end - offset)

Comment (kennytm) [reply pkg/storage/gcs.go:345]: `end + offset` is correct, see https://github.com/golang/go/blob/21a04e33353316635b5f3351e807916f3bb1e844/src/io/example_test.go#L202.

**Comment by kennytm** [comment]
the new GCS test failed in CI.

File: pkg/storage/gcs.go (hunk 1 above)
Comment (Leavrth) [reply pkg/storage/gcs.go:345]: Get it.

File: pkg/storage/gcs.go [hunk 2]
```diff
@@ -258,3 +298,82 @@ func hasSSTFiles(ctx context.Context, bucket *storage.BucketHandle, prefix strin
...
+	case io.SeekCurrent:
+		realOffset = r.pos + offset
+		if r.pos < 0 && realOffset >= 0 {
+			return 0, errors.Annotatef(berrors.ErrInvalidArgument, "Seek: offset '%v' out of range. current pos is '%v'.", offset, r.pos)
+		}
+	case io.SeekEnd:
+		if offset >= 0 {
+			return 0, errors.Annotatef(berrors.ErrInvalidArgument, "Seek: offset '%v' should be negative.", offset)
+		}
+		realOffset = offset
```
Comment (glorv) [inline pkg/storage/gcs.go:359]: ?

Comment (Leavrth) [reply pkg/storage/gcs.go:359]: see `Example(LastNBytes)` in https://pkg.go.dev/cloud.google.com/go/storage#ObjectHandle.NewRangeReader
this package support negative offset

**Comment by Leavrth** [comment]
It seems fake-gcs-server don't support negative offset yet.
https://github.com/fsouza/fake-gcs-server/blob/5a7dc5a5a672ab1b61ed7a9d4180614d60245975/fakestorage/object.go#L578-L599

**Review by glorv — APPROVED** [review]

File: pkg/storage/gcs.go [hunk 3]
```diff
@@ -155,8 +170,33 @@ func (s *gcsStorage) Open(ctx context.Context, path string) (ExternalFileReader,
...
-	panic("Unsupported Operation")
+	if opt == nil {
+		opt = &WalkOption{}
+	}
+	prefix := path.Join(s.gcs.Prefix, opt.SubDir)
+	if len(prefix) > 0 && !strings.HasSuffix(prefix, "/") {
+		prefix += "/"
+	}
+
+	query := &storage.Query{Prefix: prefix}
```
Comment (kennytm) [inline pkg/storage/gcs.go:181]: consider doing https://pkg.go.dev/cloud.google.com/go/storage#Query.SetAttrSelection to the query.

File: pkg/storage/gcs.go (hunk 2 above)
Comment (kennytm) [reply pkg/storage/gcs.go:359]: @Leavrth please include the comment https://github.com/pingcap/br/pull/1353/files#r670333536 in the code, as this behavior is quite unintuitive.

**Review by kennytm — COMMENTED** [review]
rest LGTM

File: pkg/storage/gcs.go (hunk 3 above)
Comment (Leavrth) [reply pkg/storage/gcs.go:187]: done.

File: pkg/storage/gcs.go (hunk 2 above)
Comment (Leavrth) [reply pkg/storage/gcs.go:368]: done.

**Review by kennytm — APPROVED** [review]

**Pull request merged**